pvp_core.go

```diff
@@ -100,56 +86,110 @@ func calculateRanks(stats PokemonStats, cpCap int, lvCap float64) ([16][16][16]R
 					continue
 				}
 				combinations[a][d][s] = currentStat
-				sortedRanks[c] = currentStat
+				sortedRanks[c] = &currentStat
 				c++
 			}
 		}
 	}
 
-	sort.Sort(rankingSortable(sortedRanks[:]))
+	sort.SliceStable(sortedRanks, func(i, j int) bool {
+		return comparator(sortedRanks[i], sortedRanks[j]) < 0
+	})
 
 	best := sortedRanks[0].Value
 	var i, j int16
 	for i, j = 0, 0; i < int16(len(sortedRanks)); i++ {
-		entry := &sortedRanks[i]
-		combinationsEntry := &combinations[entry.Attack][entry.Defense][entry.Stamina]
-
+		entry := sortedRanks[i]
 		percentage := roundFloat(entry.Value/best, 5)
-
 		entry.Percentage = percentage
-		combinationsEntry.Percentage = percentage
-
 		if entry.Value < sortedRanks[j].Value {
 			j = i
 		}
 		rank := j + 1
-
 		entry.Rank = rank
-		combinationsEntry.Rank = rank
 	}
 	return combinations, sortedRanks
 }
+*/
+
+func RankingComparator_Default(a, b *PvPRankingStats) int {
+	d := b.Value - a.Value
+	if d > 0 {
+		return 1
+	}
+	if d < 0 {
+		return -1
+	}
+	d = b.Attack - a.Attack
+	if d > 0 {
+		return 1
+	}
+	if d < 0 {
+		return -1
+	}
+	return 0
+}
+func RankingComparator_PreferHigherCp(a, b *PvPRankingStats) int {
+	d := RankingComparator_Default(a, b)
+	if d > 0 {
+		return 1
+	}
+	if d < 0 {
+		return -1
+	}
+	d = b.Cp - a.Cp
+	if d > 0 {
+		return 1
+	}
+	if d < 0 {
+		return -1
+	}
+	return 0
+}
+func RankingComparator_PreferLowerCp(a, b *PvPRankingStats) int {
+	d := RankingComparator_Default(a, b)
+	if d > 0 {
+		return 1
+	}
+	if d < 0 {
+		return -1
+	}
+	d = a.Cp - b.Cp
+	if d > 0 {
+		return 1
+	}
+	if d < 0 {
+		return -1
+	}
+	return 0
+}
 
 // calculateRanksCompact is optimized (for cache) core method used to calculate PvP ranks for provided Pokemon data.
-func calculateRanksCompact(stats PokemonStats, cpCap int, lvCap float64, ivFloor int) ([4096]int16, [4096]Ranking) {
-	var combinations [4096]int16
-	var sortedRanks [4096]Ranking
+func calculateRanksCompact(stats *PokemonStats, cpCap int, lvCap float64, comparator RankingComparator, ivFloor int) (*[4096]int16, *[4096]PvPRankingStats) {
+	combinations := new([4096]int16)
+	sortedRanks := new([4096]PvPRankingStats)
 
+	count := 0
 	for a := ivFloor; a <= 15; a++ {
 		for d := ivFloor; d <= 15; d++ {
 			for s := ivFloor; s <= 15; s++ {
-				entry, _ := calculatePvPStat(stats, a, d, s, cpCap, lvCap, 1)
-				entry.Index = (a*16+d)*16 + s
-				sortedRanks[entry.Index] = entry
+				if calculatePvPStat(&sortedRanks[count], stats, a, d, s, cpCap, lvCap, 1) == nil {
+					sortedRanks[count].Index = (a*16+d)*16 + s
+					count++
+				}
 			}
 		}
 	}
 
-	sort.Sort(rankingSortableIndexed(sortedRanks[:]))
+	// use Slice over SliceStable for performance since we have index available to us
+	sort.Slice(sortedRanks[:count], func(i, j int) bool {
+		d := comparator(&sortedRanks[i], &sortedRanks[j])
+		return d < 0 || d == 0 && sortedRanks[i].Index < sortedRanks[j].Index
+	})
```

Comment:
I've used `Sort` as `Slice` was noticeably slower in my benchmarks.

```
$ git checkout refined-rankings ; go build ; go test -benchtime=60s -bench=BenchmarkQueryPvPRank ; git checkout master ; go build ; go test -benchtime=60s -bench=BenchmarkQueryPvPRank
BenchmarkQueryPvPRank-24                   30672           2367168 ns/op # master
BenchmarkQueryPvPRank-24                   26839           2691408 ns/op # refined-rankings
```

`refined-rankings` is 12% slower than `master`, it's not much as we hit cache very fast, but experiment a bit with sorting, you might get better results.

Reply:
If `Slice` was slower than `Sort`, then it is a golang issue, not a me issue?

Reply:
Also according to your code, isn't `master` in fact 12% slower than `refined-rankings`?

Reply:
Yes, could be. Give it a try and run a bench when you'll get a moment.

Reply:
> Also according to your code, isn't `master` in fact 12% slower than `refined-rankings`?

[Nope.](https://www.practical-go-lessons.com/chap-34-benchmarks#how-to-read-benchmark-results)

Reply:
https://github.com/golang/go/issues/22504#issuecomment-340685108

Golang sucks so bad. How is "improving documentations" not actionable?

Reply:
Also created a PR at golang. https://github.com/golang/go/pull/59449